File: 0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        int minimum = nums.length-1;
        int[] dp = new int[nums.length];
        Arrays.fill(dp, nums.length); 
        dp[nums.length-1] = 0;

        for (int i = nums.length-2; i >=0; i--){
            int currMin = nums.length;
            for (int j = 1; j <= nums[i]; j++){
                if (i+j > nums.length-1) {
                    break;
                }
                currMin = Math.min(currMin, dp[i+j]);
            }
            dp[i] = currMin+1;
        }

        // System.out.println(Arrays.toString(dp));

        return dp[0];
    }
}

File: 0045-jump-game-ii/0045-jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        memo = [1001] * n
        memo[-1] = 0
        for i in range(n-2, -1, -1): 
            count = nums[i]
            jump = 1 if count else 1001
            memo[i] = jump + min( memo[i+1:i+1+count] if memo[i+1:i+1+count] else [0] )
        print(memo)
        return memo[0]